sql/pgwire: TestParseClientProvidedSessionParameters failed

sql/pgwire.TestParseClientProvidedSessionParameters [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8570786?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8570786?buildTab=artifacts#/) on master @ [cebffa1eb4127c2ae9bd362b462956167c732a18](https://github.com/cockroachdb/cockroach/commits/cebffa1eb4127c2ae9bd362b462956167c732a18):


```
=== RUN   TestParseClientProvidedSessionParameters
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/28616e575ae934b92eba220441c7ca42/logTestParseClientProvidedSessionParameters795899212
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestParseClientProvidedSessionParameters
    conn_test.go:1603: -- test log scope end --
    conn_test.go:1603: Leaked goroutine: goroutine 9211248 [IO wait]:
        internal/poll.runtime_pollWait(0x7f1bab75aa88, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc01d85c000?, 0xc01dca6030?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc01d85c000, {0xc01dca6030, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc01d85c000, {0xc01dca6030?, 0xc0142d5e48?, 0x47d5f2?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc01a58a000, {0xc01dca6030?, 0xc01dcc01e0?, 0xc01dcc01e0?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc005cfe160, {0x623bc78?, 0xc01dcc01e0})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
        Leaked goroutine: goroutine 9211311 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
--- FAIL: TestParseClientProvidedSessionParameters (7.49s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-sessions @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestParseClientProvidedSessionParameters.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-24151

--- a/pkg/sql/pgwire/conn_test.go
+++ b/pkg/sql/pgwire/conn_test.go
@@ -1593,6 +1593,14 @@ func TestParseClientProvidedSessionParameters(t *testing.T) {
 				_ = c.Ping(ctx)
 				// closing connection immediately, since getSessionArgs is blocking
 				_ = c.Close(ctx)
+
+				select {
+				case <-c.PgConn().CleanupDone():
+				case <-time.After(20 * time.Second):
+					// 20 seconds was picked because pgconn has an internal deadline of
+					// 15 seconds when performing the async close request.
+					t.Error("pgconn asyncClose did not clean up on time")
+				}
 			}(tc.query)
 			// Wait for the client to connect and perform the handshake.
 			_, args, err := getSessionArgs(ln, true /* trustRemoteAddr */)

